sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala

```diff
@@ -45,16 +45,23 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
    * An execution strategy for rules that indicates the maximum number of executions. If the
    * execution reaches fix point (i.e. converge) before maxIterations, it will stop.
    */
-  abstract class Strategy { def maxIterations: Int }
+  abstract class Strategy {
+
+    /** The maximum number of executions. */
+    def maxIterations: Int
+
+    /** Whether to throw exception when exceeding the maximum number. */
+    def errorOnExceed: Boolean
```

Comment:
Why did you add this in not `RuleExecutor` but `Strategy`?

Reply:
> Why did you add this in not `RuleExecutor` but `Strategy`?

@maropu The setting `SQLConf.ANALYZER_MAX_ITERATIONS` only controls a subset of Analyzer Rule batches (Hints, Substitution, Resolution and Cleanup). For other Rules with `Once` strategy, the original RuleExecutor logic will not raise exception or log warning. Which means within Analyzer, it will take different actions for different Batches/Strategy. So I think the `errorOnExceed` behavior is more like binding with maxIteration in Strategy. By doing this, we could have finer granularity to support the feature and provide better flexibility to control behaviors for different batches in the future. Thanks!

Reply:
Ur, I see. Thanks. Since this method is used only once now in `AnalysisSuite`, can you move this into `AnaysisSuite`?

Reply:
> Ur, I see. Thanks. Since this method is used only once now in `AnalysisSuite`, can you move this into `AnaysisSuite`?

Sure, will do :-)

Reply:
(Oh, I noticed now that I left the comment in the wrong position, sorry...)